File: day4/python/main.py
class MyCircularQueue:
    def __init__(self, k: int):
        self.data = [0] * k
        self.front = 0
        self.rear = 0
        self.size = 0
        self.capacity = k

    def enQueue(self, value: int) -> bool:
        if self.isFull(): return False
        self.data[self.rear] = value
        self.rear = (self.rear + 1) % self.capacity
        self.size += 1
        return True

    def deQueue(self) -> bool:
        if self.isEmpty(): return False
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return True

    def Front(self) -> int:
        return -1 if self.isEmpty() else self.data[self.front]

    def Rear(self) -> int:
        return -1 if self.isEmpty() else self.data[(self.rear - 1 + self.capacity) % self.capacity]

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size == self.capacity

File: day4/cpp/main.cpp
#include <vector>

class MyCircularQueue
{
private:
    std::vector<int> data;
    int front, rear, size, capacity;

public:
    MyCircularQueue(int k) : data(k), front(0), rear(0), size(0), capacity(k) {}

    bool enQueue(int value)
    {
        if (isFull())
            return false;
        data[rear] = value;
        rear = (rear + 1) % capacity;
        ++size;
        return true;
    }

    bool deQueue()
    {
        if (isEmpty())
            return false;
        front = (front + 1) % capacity;
        --size;
        return true;
    }

    int Front()
    {
        return isEmpty() ? -1 : data[front];
    }

    int Rear()
    {
        return isEmpty() ? -1 : data[(rear + capacity - 1) % capacity];
    }

    bool isEmpty()
    {
        return size == 0;
    }

    bool isFull()
    {
        return size == capacity;
    }
};
